File: core/src/test/scala/diplomacy/variant/standard/rule/RetreatValidatorSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.rule.InvalidOrderMessage
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board._

class RetreatVailidatorSpec extends UnitSpec {
  val validator = new RetreatValidator[T, map.Power]
  val board =
    Board[State[T], map.Power, MilitaryBranch, UnitStatus, ProvinceStatus](
      map.map,
      1901.Spring - Retreat,
      Set(
        DiplomacyUnit(France, Army, Bur),
        DiplomacyUnit(France, Army, Mar),
        DiplomacyUnit(Italy, Fleet, Wes)
      ), Map(),
      Map(
        DiplomacyUnit(France, Army, Mar) -> UnitStatus.Dislodged(Gas.province),
        DiplomacyUnit(Italy, Fleet, Wes) -> UnitStatus.Dislodged(Tyn.province)
      ),
      Map(Pie.province -> ProvinceStatus.Standoff)
    )
  val $ = StandardRuleOrderHelper(board)
  import $._

  "A retreat-validator" when {
    "a valid order is received" should {
      "return null (1)." in {
        validator.errorMessageOfOrder(board)(A(Mar).retreat(Spa)) should be(None)
      }
      "return null (2)." in {
        validator.errorMessageOfOrder(board)(A(Mar).disband()) should be(None)
      }
    }
    "an order that its target unit is not dislodged is received" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(Order.Disband(DiplomacyUnit(France, Army, Bur))) should be(
          Option(InvalidOrderMessage("A Bur is not dislodged."))
        )
      }
    }
    "an order that its target unit retreat to an invalid location" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(A(Mar).retreat(Gas)) should be(
          Option(InvalidOrderMessage("A Mar cannot retreat to Gas."))
        )
      }
    }

    "a set or orders that some dislodged unit have no order" should {
      "return an error message." in {
        validator.errorMessageOfOrders(board)(Set(A(Mar).retreat(Gas))) should be(
          Option(InvalidOrderMessage("F Wes has no order."))
        )
      }
    }
  }
}

File: test/variant/standard/rule/retreat-validator-spec.ts
import * as chai from "chai"
import {
  RetreatValidator, Board, Unit, MilitaryBranch, State, Result, Phase, ProvinceStatus, Dislodged,
  StandardRuleHelper as Helper
} from "../../../../src/variant/standard/rule/module"
import * as Order from "./../../../../src/variant/standard/rule/order"
import * as Error from "./../../../../src/variant/standard/rule/error"
import { Executed } from "./../../../../src/rule/module"
import { locations as $, Power, map } from "./../../../../src/variant/standard/map/module"
import { Season, Turn } from "./../../../../src/variant/standard/board/module"

const { Army, Fleet } = MilitaryBranch
const { Retreat } = Phase

const should = chai.should()

const validator = new RetreatValidator()

describe("A RetreatValidator", () => {
  const u1 = new Unit(Army, $.Bur, Power.France)
  const u2 = new Unit(Army, $.Mar, Power.France)
  const u3 = new Unit(Fleet, $.Wes, Power.Italy)
  const board = new Board<Power>(
    map, new State(new Turn(1901, Season.Autumn), Retreat),
    [u1, u2, u3],
    [
      [u2, new Dislodged($.Gas.province)], [u3, new Dislodged($.Tyn.province)]
    ],
    [[$.Pie.province, new ProvinceStatus(null, true)]]
  )
  const $$ = new Helper(board)
  describe("when a valid order is received", () => {
    it("returns null (1)", () => {
      should.equal(
        validator.errorMessageOfOrder(board, $$.A($.Mar).retreat($.Spa)), null
      )
    })
    it("returns null (2)", () => {
      should.equal(
        validator.errorMessageOfOrder(board, $$.A($.Mar).disband()), null
      )
    })
    describe("an order that its target is not dislodged is received", () => {
      it("returns an error message", () => {
        new Error.CannotBeOrdered(new Order.Disband(u1)).should.deep.equal(
          validator.errorMessageOfOrder(board, new Order.Disband(u1))
        )
      })
    })
    describe("an order that its target retreats to an invalid location", () => {
      it("returns an error message", () => {
        new Error.UnmovableLocation($$.A($.Mar).unit, $.Gas).should.deep.equal(
          validator.errorMessageOfOrder(board, $$.A($.Mar).retreat($.Gas))
        )
      })
    })
    describe("a set of orders that some dislodged unit have no order", () => {
      it("returns an error message", () => {
        new Error.OrderNotExisted($$.F($.Wes).unit).should.deep.equal(
          validator.errorMessageOfOrders(board, new Set([$$.A($.Mar).retreat($.Gas)]))
        )
      })
    })
  })
})
